sound/soc/sof/ipc4-loader.c

```diff
@@ -60,67 +60,110 @@ static size_t sof_ipc4_fw_parse_ext_man(struct snd_sof_dev *sdev)
-						   sizeof(*fw_module), GFP_KERNEL);
-	if (!ipc4_data->fw_modules)
-		return -ENOMEM;
+	/*
+	 * allocate memory for the number of modules in the base FW + the max libraries
+	 * supported by the firmware
+	 */
+	if (!lib_index) {
+		/* max_fw_libs includes the base firmware */
+		int num_total_modules = fw_header->num_module_entries + ipc4_data->max_fw_libs - 1;
+
+		ipc4_data->fw_modules = devm_kmalloc_array(sdev->dev, num_total_modules,
+							   sizeof(*fw_module), GFP_KERNEL);
+		if (!ipc4_data->fw_modules)
+			return -ENOMEM;
+
+		ipc4_data->module_uuids = devm_kcalloc(sdev->dev, num_total_modules,
+						       sizeof(guid_t), GFP_KERNEL);
+		if (!ipc4_data->module_uuids)
+			return -ENOMEM;
+	} else {
+		char *fw_name = kasprintf(GFP_KERNEL, "%s", fw_header->name);
+
+		if (!fw_name)
+			return -ENOMEM;
+
+		/* set the FW name for the 3rd party modules */
+		ipc4_data->module_lib_info[lib_index].name = fw_name;
+	}
 
-	ipc4_data->num_fw_modules = fw_header->num_module_entries;
-	fw_module = ipc4_data->fw_modules;
+	fw_module = ipc4_data->fw_modules + ipc4_data->num_fw_modules;
 
 	fm_entry = (struct sof_man4_module *)((u8 *)fw_header + fw_header->len);
 	remaining -= fw_header->len;
 
 	if (remaining < fw_header->num_module_entries * sizeof(*fm_entry)) {
 		dev_err(sdev->dev, "Invalid num_module_entries %u\n",
 			fw_header->num_module_entries);
-		return -EINVAL;
+		ret = -EINVAL;
+		goto err;
 	}
 
 	fm_config = (struct sof_man4_module_config *)
 				(fm_entry + fw_header->num_module_entries);
 	remaining -= (fw_header->num_module_entries * sizeof(*fm_entry));
 	for (i = 0; i < fw_header->num_module_entries; i++) {
-		memcpy(&fw_module->man4_module_entry, fm_entry, sizeof(*fm_entry));
+		/* ignore modules with null UUID */
+		if (guid_is_null(&fm_entry->uuid)) {
+			fm_entry++;
+			continue;
+		}
 
 		if (fm_entry->cfg_count) {
 			if (remaining < (fm_entry->cfg_offset + fm_entry->cfg_count) *
 			    sizeof(*fm_config)) {
 				dev_err(sdev->dev, "Invalid module cfg_offset %u\n",
 					fm_entry->cfg_offset);
-				return -EINVAL;
+				ret = -EINVAL;
+				goto err;
 			}
 
 			/* a module's config is always the same size */
 			fw_module->bss_size = fm_config[fm_entry->cfg_offset].is_bytes;
 
+			guid_copy(&ipc4_data->module_uuids[i + ipc4_data->num_fw_modules],
+				  &fm_entry->uuid);
+
 			dev_dbg(sdev->dev,
 				"module %s: UUID %pUL cfg_count: %u, bss_size: %#x\n",
-				fm_entry->name, &fm_entry->uuid, fm_entry->cfg_count,
-				fw_module->bss_size);
+				fm_entry->name,
+				&ipc4_data->module_uuids[i + ipc4_data->num_fw_modules],
+				fm_entry->cfg_count, fw_module->bss_size);
 		} else {
 			fw_module->bss_size = 0;
 
 			dev_dbg(sdev->dev, "module %s: UUID %pUL\n", fm_entry->name,
 				&fm_entry->uuid);
 		}
 
-		fw_module->man4_module_entry.id = i;
+		memcpy(&fw_module->man4_module_entry, fm_entry, sizeof(*fm_entry));
+
+		fw_module->man4_module_entry.id = i + ipc4_data->num_fw_modules;
```

Comment:
For basefw: man4_module_entry.id = 0...num_fw_modules_in_base_fw-1,
for the first library loaded: man4_module_entry.id = 0x1000 + 0...num_fw_modules_in_the_library-1
for the second library loaded: man4_module_entry.id = 0x2000 + 0...num_fw_modules_in_the_library-1
...
@libinyang found this from pytest.

I think you are merging modules from all libraries, this is easy to implement, but I am thinking something like this, I think it is module-unload friendly.
![image](https://user-images.githubusercontent.com/52959957/175846071-8171704c-0a47-4cf6-90fd-469f5dd7148b.png)

Reply:
@aiChaoSONG I'd argue there is value in keeping the modules separate per library. When would we have a library that contains more than one module? If we're tying the widget UUID to the library name, this can never really work

Reply:
I did some test for you reference. There are two modules for some modules

````
rander@cml-s:/media/xxx/adsp_fw_parser$ ./adsp_fw_parser ~/fw_libs/tgl_maxx_audio_pro.bin | grep uuid
                    "uuid": "52983D04-2414-88B4-A2A2-C1397E13B022",
                    "uuid": "FAACC8CC-B365-4964-B4B8-BD4DEB18D922",

rander@cml-s:/media/xxx/adsp_fw_parser$ ./adsp_fw_parser ~/fw_libs/cml_maxx_audio_pro.bin | grep uuid
                    "uuid": "52983D04-2414-88B4-A2A2-C1397E13B022",
                    "uuid": "FAACC8CC-B365-4964-B4B8-BD4DEB18D922",
````

Reply:
> I think you are merging modules from all libraries, this is easy to implement, but I am thinking something like this, I think it is module-unload friendly.

module unload is not supported to the best of my knowledge @aiChaoSONG?

Reply:
On option might be (and probably ugly) is to use symlinks when a binary file contains multiple modules with different UUID?
```
ls -al
52983D04-2414-88B4-A2A2-C1397E13B022.bin -> tgl_maxx_audio_pro.bin
FAACC8CC-B365-4964-B4B8-BD4DEB18D922.bin -> tgl_maxx_audio_pro.bin
tgl_maxx_audio_pro.bin
```
and load only the requested module from the loaded firmware file?

Reply:
or we load all modules form the file for simplicity.

Reply:
we can have a config file or environment variable (different platform may has different library) to set library path and load these libraries at first boot up. On windows platform, there is a library path defined in registry and driver loads all libraries in boot-up stage